### PR TITLE
Merge s4642448's Siamese Network Implementation

### PR DESCRIPTION
**Pull request by nordkamp**
Afternoon Shakes, I'm looking to merge my Siamese network implementation. Thanks!

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **15**

Solves problem (**2**): The solution is appropriate for the problem but there is no classifier and no performance to evaluate the model.

Implementation functions as intended: Not quite (**1**)

Good design: good-designed (**1**)

Commenting: sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

**Note**:
1. not data split into training, validation, and test sets or using k-fold cross validation training
2. The evaluation lacks a standardized accuracy measure and formal testing loop, stopping at TARGET_SAMPLES and potentially creating biased, incomplete results.
3. Accuracy is calculated within each batch instead of at the end, which can lead to fluctuating metrics instead of a cumulative understanding of performance.
4. The extract_loss function filters and logs loss values, but there is no clear integration with validation loss or early stopping. This limits understanding of the model's performance.
5. The code seems to be primarily a verification-style model (pair-based evaluation) rather than a direct classifier for benign/malignant prediction. (For benign/malignant classification, you typically need a classifier that directly assigns a label (benign or malignant) to individual images. The current pair-based approach can be adapted by first training the Siamese network to learn feature representations and then using the learned embeddings as input to a separate classifier that outputs the benign/malignant labels. )

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>No design and implementation. </td><td>-2</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. No classifier, loss function is incorrect.</td><td>-4</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>Data leakage found. No data splits.</td><td>-2</td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. </td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-2</td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-16</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Matthew 	Hoffman

**Comment by shakes76** [comment]
Model is incomplete, partially solution. -2 additionally.
